win32ss/user/ntuser/message.c

```diff
@@ -483,32 +483,26 @@ CopyMsgToKernelMem(MSG *KernelModeMsg, MSG *UserModeMsg, PMSGMEMORY MsgMemoryEnt
         {
             TRACE("Copy Message %u from usermode buffer\n", KernelModeMsg->message);
             /* Don't do extra testing for 1 word messages. For examples see
-             * https://wiki.winehq.org/List_Of_Windows_Messages. */
-            if (Size > 1)
+             * https://wiki.winehq.org/List_Of_Windows_Messages and
+             * we are just handling WM_WININICHANGE here. */
+            if (Size > 1 && UserModeMsg->lParam &&
+                KernelModeMsg->message == WM_WININICHANGE)
             {
                 WCHAR lParamMsg[_countof(StrUserKernel[0]) + 1] = { 0 };
                 _SEH2_TRY
                 {
-                    if (UserModeMsg->lParam)
-                        RtlCopyMemory(lParamMsg, (WCHAR*)UserModeMsg->lParam, sizeof(lParamMsg));
-                        /* Make sure that the last WCHAR is a UNICODE_NULL */
-                        for (i = 0; i < ARRAYSIZE(lParamMsg); ++i)
-                        {
-                            if (lParamMsg[i] == 0)
-                                break;
-                        }
-                        /* If we did not find a UNICODE_NULL, then set last WCHAR to one */
-                        if (i == ARRAYSIZE(lParamMsg))
-                            lParamMsg[_countof(StrUserKernel[0])] = UNICODE_NULL;
+                    RtlCopyMemory(lParamMsg, (WCHAR*)UserModeMsg->lParam, sizeof(lParamMsg));
                 }
                 _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
                 {
                     _SEH2_YIELD(return STATUS_ACCESS_VIOLATION);
                 }
                 _SEH2_END;
 
-                if (UserModeMsg->lParam && !UserModeMsg->wParam &&
-                    PosInArray(lParamMsg) >= 0)
+                /* Make sure that we have a UNICODE_NULL within lParamMsg */
+                lParamMsg[ARRAYSIZE(lParamMsg) - 1] = UNICODE_NULL;
```

Comment:
Repeating my question from here: https://github.com/reactos/reactos/pull/7215#discussion_r1712621963
Do you have any reason to favor ARRAY_SIZE() over count_of()? This is not a request for change yet. But a question with the intention to learn something maybe.

Reply:
They both have their own downsides.

 - _countof is technically a reserved identifier in the global scope.
 - In C++, ARRAYSIZE can be a template (in the MS SDK) and this template version is broken on some things where the C macro works correctly (winnt.h provides RTL_NUMBER_OF_V1 and RTL_NUMBER_OF_V2 for this reason).